### PR TITLE
5035/exception code

### PR DESCRIPTION
**Pull request by luu-alex**
#5035 
## Description
From this [feature request](https://github.com/web3/web3.js/issues/5035) An Error that happens in the rpc will now create an rpc error which will include the error code, message and jsonrpc. 
It will include standard error codes and unstandard error codes from [eip 1747](https://eips.ethereum.org/EIPS/eip-1474)

Code | Message | Meaning | Category
-- | -- | -- | --
-32700 | Parse error | Invalid JSON | standard
-32600 | Invalid request | JSON is not a valid request object | standard
-32601 | Method not found | Method does not exist | standard
-32602 | Invalid params | Invalid method parameters | standard
-32603 | Internal error | Internal JSON-RPC error | standard
-32000 | Invalid input | Missing or invalid parameters | non-standard
-32001 | Resource not found | Requested resource not found | non-standard
-32002 | Resource unavailable | Requested resource not available | non-standard
-32003 | Transaction rejected | Transaction creation failed | non-standard
-32004 | Method not supported | Method is not implemented | non-standard
-32005 | Limit exceeded | Request exceeds defined limit | non-standard
-32006 | JSON-RPC version not supported | Version of JSON-RPC protocol is not supported | non-standard

<br class="Apple-interchange-newline">

Please include a summary of the changes and be sure to follow our [Contribution Guidelines](../blob/1.x/CONTRIBUTIONS.md).

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist for 1.x:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [ ] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` with success.
- [ ] I have tested the built `dist/web3.min.js` in a browser.
- [ ] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [ ] I have updated the `CHANGELOG.md` file in the root folder.

## Checklist for 4.x:

- [ ] I have selected the correct 4.x base branch.
- [ ] Within the description, the feature or issue is discussed in detail or an issue is linked.   
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added any required tests for the changes I made 
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran `yarn` successfully
- [ ] I ran `yarn lint` successfully
- [ ] I ran `yarn build:web` successfully
- [ ] I ran `yarn test:unit` successfully
- [ ] I ran `yarn test:integration` successfully
- [ ] I ran `compile:contracts` successfully
- [ ] I have tested my code.
- [ ] I have updated the corresponding `CHANGELOG.md` file in the packages I have edited.

### CODE REVIEW

**Comment by luu-alex** [comment]
Non-standard error codes may confuse users if a node doesn't use the error-codes given by eip 1747. 
For example using local ganache, the error code `-32000` is `method not supported `when in the eip 1747 it is considered a `Invalid input`.
I believe we can put a check on both the errorcode and the error message from the node to make sure that an rpc error is truly following eip 1747. Thoughts?

File: packages/web3-errors/src/errors/rpc_errors.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	'-32003': {
+		message: 'Transaction rejected',
+	},
+	'-32004': {
+		message: 'Method not supported',
+	},
+	'-32005': {
+		message: 'Limit exceeded',
+	},
+	'-32006': {
```
Comment (jdevcs) [inline packages/web3-errors/src/errors/rpc_errors.ts]: I think create dictionary/map with directly using existing error codes ( for avoiding any error code duplication and future maintenance) . 

```
    export const rpcErrorsMap = new Map<number, object>();
    rpcErrorsMap.set(ERR_RPC_INVALID_JSON, ParseError);
    rpcErrorsMap.set(ERR_RPC_INVALID_REQUEST, InvalidRequestError);
    //.....
```

File: packages/web3-utils/src/json_rpc.ts [hunk 2]
```diff
@@ -28,9 +28,71 @@ import {
...
+	MethodNotSupported,
+	LimitExceededError,
+	VersionNotSupportedError,
+	RpcError,
+	rpcErrorValues,
+} from 'web3-errors';
 import { uuidV4 } from './uuid';
 
+export const getRpcError = (code: number) => {
+	const codeStr = code.toString();
```
Comment (jdevcs) [inline packages/web3-utils/src/json_rpc.ts]: if we use map as shared above we can avoid these error codes duplication as well and can directly query that map where required as:
`rpcErrorsMap.get(-32700) ` // in web3-core request manager directly

File: packages/web3-utils/src/json_rpc.ts [hunk 3]
```diff
@@ -28,9 +28,71 @@ import {
...
+	if (codeStr === '-32700') {
+		return ParseError;
+	}
+	if (codeStr === '-32600') {
```
Comment (jdevcs) [inline packages/web3-utils/src/json_rpc.ts]: There shouldn't be multiple copies of these error codes in web3js lib multiple places.  we should use error_codes.ts.

File: packages/web3-rpc-methods/src/eth_rpc_methods.ts [hunk 4]
```diff
@@ -44,9 +44,16 @@ export async function getSyncing(requestManager: Web3RequestManager) {
...
+	return requestManager.send({
+		method: 'test',
+		params: [],
+	});
+}
+
 export async function getCoinbase(requestManager: Web3RequestManager) {
 	return requestManager.send({
-		method: 'eth_coinbase',
+		method: 'test',
```
Comment (jdevcs) [inline packages/web3-rpc-methods/src/eth_rpc_methods.ts]: this is for debugging only?

**Comment by jdevcs** [comment]
This PR lgtm, but some of tests are not passing in CI, pls check these. Thanks Alex, 
Also create issue for integration/e2e tests of these error codes.

**Comment by luu-alex** [comment]
Thank you @jdevcs, created the issue and still debugging.  Errors were happening as ganache and geth dont follow the [unstandard error codes](https://github.com/web3/web3.js/pull/5735/files#diff-65dc5d05081c69f1865f621e1549f8254caa34650e13756fcbd9a4374eaa9e98R421), they would give different error messages. For example error code `-32000` in Geth is `gas too low`, but according to eip-1474 it is `invalid input`. I created a [check](https://github.com/web3/web3.js/pull/5735/files#diff-90d726c886e913b9fbd11b83e2b4160d084b075f1ca99ab809bf1baa8a743700R330) that error code and message has to match to be an eip 1474 error.

**Comment by jdevcs** [comment]
> Thank you @jdevcs, created the issue and still debugging. Errors were happening as ganache and geth dont follow the [unstandard error codes](https://github.com/web3/web3.js/pull/5735/files#diff-65dc5d05081c69f1865f621e1549f8254caa34650e13756fcbd9a4374eaa9e98R421), they would give different error messages. For example error code `-32000` in Geth is `gas too low`, but according to eip-1474 it is `invalid input`. I created a [check](https://github.com/web3/web3.js/pull/5735/files#diff-90d726c886e913b9fbd11b83e2b4160d084b075f1ca99ab809bf1baa8a743700R330) that error code and message has to match to be an eip 1474 error.

Its because https://eips.ethereum.org/EIPS/eip-1474 EIP is not approved and its not necessary that Eth clients implement this EIP. May be add some flag ( via web3 config ) for turning on this RPC errors feature in web3, instead of keep it always on?

**Review by nikoulai — APPROVED** [review]

**Review by jdevcs — APPROVED** [review]

**Review by Aheyachy — APPROVED** [review]

**Pull request merged**